File: exercise4v1/python/vargmm.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.io import loadmat
from scipy.stats import multivariate_normal, wishart
from scipy.special import digamma
from mpl_toolkits.mplot3d import Axes3D

## Helper function for plotting a 2D Gaussian
def plot_normal(mu, Sigma):
  l, V = np.linalg.eigh(Sigma)
  l[l<0] = 0
  t = np.linspace(0.0, 2.0*np.pi, 100)
  xy = np.stack((np.cos(t), np.sin(t)))
  Txy = mu + ((V * np.sqrt(l)).dot(xy)).T
  plt.plot(Txy[:, 0], Txy[:, 1])

def logdet(Sigma):
  (s, ulogdet) = np.linalg.slogdet(Sigma)
  return s*ulogdet

## Load data
data = loadmat('clusterdata2d.mat')['data']
N, D = data.shape

## Number of components/clusters
K = 10

## Priors
alpha0 = 1e-3 # Mixing prior (small number: let the data speak)
m0 = np.zeros(D); beta0 = 1e-3 # Gaussian mean prior
v0 = 3e1; W0 = np.eye(D)/v0 # Wishart covariance prior

## Initialize parameters
m_k = []
W_k = []
beta_k  = np.repeat(beta0  + N/K, K)
alpha_k = np.repeat(alpha0 + N/K, K)
v_k     = np.repeat(v0 + N/K,     K)
for _ in range(K):
  # Let m_k be a random data point:
  m_k.append(data[np.random.choice(N)])
  # Let W_k be the mean of the Wishart prior:
  W_k.append(v0*W0)


## Loop until you're happy
max_iter = 100
ln_rho = np.zeros((N, K))
for iteration in range(max_iter):
  ## Variational E-step
  # XXX: FILL ME IN!
  
  ## Variational M-step
  # XXX: FILL ME IN!
  
## Plot data with distribution (we show expected distribution)
plt.figure(1)
plt.plot(data[:, 0], data[:, 1], '.')
for k in range(K):
  if Nk[k] > 0:
    plot_normal(m_k[k], np.linalg.pinv(v_k[k] * (W_k[k])))
plt.show()

## Now, animate the uncertainty by sampling
#num_samples = 100
#plt.figure(2)
#for s in range(num_samples):
#  plt.clf()
#  plt.plot(data[:, 0], data[:, 1], '.')
#  for k in range(K):
#    if Nk[k] > 0:
#      L = wishart.rvs(scale=W_k[k], df=v_k[k])
#      Sigma = np.linalg.pinv(L)
#      mu = multivariate_normal.rvs(mean=m_k[k], cov=Sigma/beta_k[k])
#      plot_normal(mu, Sigma)
#  plt.pause(0.01)

## Animate the entire mixture distribution
#fig = plt.figure(3)
#ax = fig.add_subplot(111, projection='3d')
#lim = np.linspace(-2, 2, 20)
#X, Y = np.meshgrid(lim, lim)
#XY = np.vstack((X.flatten(), Y.flatten())).T # 2500x2
#for s in range(num_samples):
#  pi_k = np.random.dirichlet(alpha_k)
#  Z = np.zeros(X.shape).flatten()
#  for k in range(K):
#    L = wishart.rvs(scale=W_k[k], df=v_k[k])
#    Sigma = np.linalg.pinv(L)
#    mu = multivariate_normal.rvs(mean=m_k[k], cov=Sigma/beta_k[k])
#    Z += pi_k[k] * multivariate_normal.pdf(XY, mean=mu, cov=Sigma)
#  ax.cla()
#  ax.plot_surface(X, Y, Z.reshape(X.shape))
#  ax.set_zlim(top=0.3)
#  plt.pause(0.01)

File: exercise4v1/matlab/vargmm.m
clear all
close all

%% Load data
load clusterdata2d % gives 'data'
[N, D] = size(data);

%% Wrapper for digamma function
digamma = @(x) psi(0, x);

%% Number of components/clusters
K = 10;

%% Priors
a0 = 1e-3; % Mixing prior (small number: let the data speak)
m0 = zeros(1, D); B0 = 1e-3; % Gaussian mean prior
v0 = 3e1; W0 = eye(D)/v0; % Wishart covariance prior

%% Initialize parameters
m = cell(K, 1);
W = cell(K, 1);
B = repmat(B0 + N/K, [1, K]); % 1xK
a = repmat(a0 + N/K, [1, K]); % 1xK
v = repmat(v0 + N/K, [1, K]); % 1xK
for k = 1:K
    % Let m_k be a random data point:
    m{k} = data(randi(N), :); % 1xD
    % Let W_k be the mean of the Wishart prior:
    W{k} = v0*W0; % DxD % I MODIFIED HERE TODO
end % for
Nk = ones(1, K) / K; % 1xK

%% Loop until you're happy
max_iter = 100;
xb = cell(K,1);
S = cell(K,1);
lnrho = zeros(N,K);
r = zeros(N,K);
for iter = 1:max_iter
    %% Variational E-step
    % XXX: FILL ME IN!
    Elnpb = digamma(a) - digamma(sum(a));
    for k = 1:K
        ElnLt = sum(digamma((v(k)+1-(1:D))/2)) + D*log(2) + logdet(W{k});
        EmL = D/B(k) + v(k)*sum(((data - m{k})*W{k}).*(data - m{k}),2);
        lnrho(:,k) = Elnpb(k) + ElnLt/2 - D/2*log(2*pi) - 1/2*EmL;
    end
    lnC = -max(lnrho,[],2);
    r = exp(lnrho-lnC);
    r = r./sum(r,2);
    Nk = sum(r,1);
    
    
    %% Variational M-step
    % XXX: FILL ME IN!
    for k = 1:K
        rk = r(:,k)/Nk(k);
        rk(isnan(rk)) = 0;
        xb{k} = (rk'*data);
        S{k} = (rk.*(data - xb{k}))'*(data - xb{k});
        a(k) = a0 + Nk(k);
        B(k) = B0 + Nk(k);
        m{k} = (B0*m0 + Nk(k)*xb{k})/B(k);
        W{k} = pinv(inv(W0) + Nk(k)*S{k} + B0*Nk(k)/(B0 + Nk(k))*((xb{k} - m0)'*(xb{k} - m0)));
        v(k) = v0 + Nk(k);
    end
    v = v0 + Nk;
    
end % for

%% Plot data with distribution (we show expected distribution)
figure
plot(data(:, 1), data(:, 2), '.');
hold on
for k = 1:K
  if (Nk(k) > 0)
    plot_normal(m{k}, pinv(v(k) * (W{k})), 'linewidth', 2);
  end % if
end % for
hold off

%% Now, animate the uncertainty by sampling
%{
num_samples = 100;
figure
for s = 1:num_samples
  plot(data(:, 1), data(:, 2), '.');
  hold on
  for k = 1:K
    if (Nk(k) > 0)
      L = wishrnd(W_k{k}, v_k(k));
      Sigma = pinv(L);
      mu = mvnrnd(m_k{k}, Sigma/beta_k(k));
      plot_normal(mu, Sigma, 'linewidth', 2);
    end % if
  end % for
  hold off
  pause(0.1)
end % for

%% Animate the entire mixture distribution
figure
[X, Y] = meshgrid(linspace(-2, 2, 50));
XY = [X(:), Y(:)]; % 2500x2
for s = 1:num_samples
  pi_k = dirrnd(alpha_k);
  Z = zeros(size(X));
  for k = 1:K
    L = wishrnd(W_k{k}, v_k(k));
    Sigma = pinv(L);
    mu = mvnrnd(m_k{k}, Sigma/beta_k(k));
    Z(:) = Z(:) + pi_k(k) * mvnpdf(XY, mu, Sigma);
  end % for
  h = surfl(X, Y, Z); set(h, 'edgecolor', 'none');
  axis([-2, 2, -2, 2, 0, 0.3]);
  view(200*s/num_samples, 30);
  pause(0.1);
end % for
%}
